File: seabattle.js
// seabattle.js
#!/usr/bin/env node
'use strict';

const readline = require('readline');
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

const COLORS = {
    reset: '\x1b[0m',
    blue: '\x1b[94m',
    green: '\x1b[92m',
    red: '\x1b[91m',
    yellow: '\x1b[93m',
    cyan: '\x1b[96m',
    gray: '\x1b[90m',
    bold: '\x1b[1m'
};

function colorize(text, color) {
    return COLORS[color] + text + COLORS.reset;
}

const SIZE = 10;
const SHIP_LENGTHS = [4,3,3,2,2,2,1,1,1,1];

class SeaBattle {
    constructor(mode, difficulty, manual) {
        this.mode = mode;
        this.difficulty = difficulty;
        this.manual = manual;
        this.board1 = Array.from({length:SIZE}, () => Array(SIZE).fill('~'));
        this.board2 = Array.from({length:SIZE}, () => Array(SIZE).fill('~'));
        this.shot1 = Array.from({length:SIZE}, () => Array(SIZE).fill('~'));
        this.shot2 = Array.from({length:SIZE}, () => Array(SIZE).fill('~'));
        this.ships1 = [];
        this.ships2 = [];
        this.alive1 = [];
        this.alive2 = [];
        this.shots = 0;
        this.hits = 0;
        this.turn = 0;
    }

    parseCoord(s) {
        const m = s.match(/^([A-Ja-j])(\d{1,2})$/);
        if (!m) return null;
        const col = m[1].toUpperCase().charCodeAt(0) - 65;
        const row = parseInt(m[2]) - 1;
        if (row < 0 || row >= SIZE || col < 0 || col >= SIZE) return null;
        return [row, col];
    }

    coordStr(r,c) {
        return String.fromCharCode(65+c) + (r+1);
    }

    canPlace(board, row, col, length, dir) {
        for (let i=0; i<length; i++) {
            let r = dir==='v' ? row+i : row;
            let c = dir==='h' ? col+i : col;
            if (r>=SIZE || c>=SIZE || board[r][c] !== '~') return false;
            for (let dr=-1; dr<=1; dr++) {
                for (let dc=-1; dc<=1; dc++) {
                    let nr=r+dr, nc=c+dc;
                    if (nr>=0 && nr<SIZE && nc>=0 && nc<SIZE && board[nr][nc] !== '~' && board[nr][nc] !== '#') {
                        if (board[nr][nc] === 'S' || board[nr][nc] === 'X') return false;
                    }
                }
            }
        }
        return true;
    }

    placeShip(board, row, col, length, dir) {
        for (let i=0; i<length; i++) {
            let r = dir==='v' ? row+i : row;
            let c = dir==='h' ? col+i : col;
            board[r][c] = 'S';
        }
    }

    placeRandom(board, ships) {
        for (const length of SHIP_LENGTHS) {
            let placed = false;
            let attempts = 0;
            while (!placed && attempts < 1000) {
                attempts++;
                let row = Math.floor(Math.random()*SIZE);
                let col = Math.floor(Math.random()*SIZE);
                let dir = Math.random()<0.5 ? 'h' : 'v';
                if (this.canPlace(board, row, col, length, dir)) {
                    this.placeShip(board, row, col, length, dir);
                    let coords = [];
                    for (let i=0; i<length; i++) {
                        let r = dir==='v' ? row+i : row;
                        let c = dir==='h' ? col+i : col;
                        coords.push([r,c]);
                    }
                    ships.push(coords);
                    placed = true;
                }
            }
            if (!placed) {
                for (let i=0; i<SIZE; i++) board[i].fill('~');
                ships.length = 0;
                this.placeRandom(board, ships);
                return;
            }
        }
    }

    manualPlace(board, ships) {
        return new Promise((resolve) => {
            console.log('Ручная расстановка кораблей.');
            const placeNext = (idx) => {
                if (idx >= SHIP_LENGTHS.length) { resolve(); return; }
                const length = SHIP_LENGTHS[idx];
                console.log(`Разместите корабль длины ${length}`);
                this.printBoard(board, true);
                rl.question('Введите координату (A1): ', (coord) => {
                    if (coord === 'q') process.exit();
                    rl.question('Введите направление (h/v): ', (dir) => {
                        const pos = this.parseCoord(coord);
                        if (!pos || (dir !== 'h' && dir !== 'v')) {
                            console.log('Неверный ввод.');
                            placeNext(idx);
                            return;
                        }
                        const [row, col] = pos;
                        if (this.canPlace(board, row, col, length, dir)) {
                            this.placeShip(board, row, col, length, dir);
                            let coords = [];
                            for (let i=0; i<length; i++) {
                                let r = dir==='v' ? row+i : row;
                                let c = dir==='h' ? col+i : col;
                                coords.push([r,c]);
                            }
                            ships.push(coords);
                            console.log('Корабль размещён.');
                            placeNext(idx+1);
                        } else {
                            console.log('Нельзя разместить здесь.');
                            placeNext(idx);
                        }
                    });
                });
            };
            placeNext(0);
        });
    }

    printBoard(board, showShips=false) {
        process.stdout.write('   ');
        for (let i=0; i<SIZE; i++) process.stdout.write(String.fromCharCode(65+i) + ' ');
        console.log();
        for (let r=0; r<SIZE; r++) {
            process.stdout.write((r+1).toString().padStart(2) + ' ');
            for (let c=0; c<SIZE; c++) {
                const cell = board[r][c];
                if (cell === '~') process.stdout.write(colorize('· ', 'gray'));
                else if (cell === 'S' && showShips) process.stdout.write(colorize('S ', 'green'));
                else if (cell === 'X') process.stdout.write(colorize('X ', 'red'));
                else if (cell === 'O') process.stdout.write(colorize('O ', 'yellow'));
                else process.stdout.write(cell + ' ');
            }
            console.log();
        }
    }

    makeShot(row, col, alive, shotBoard, enemyBoard) {
        if (enemyBoard[row][col] === 'S') {
            const idx = alive.findIndex(p => p[0]===row && p[1]===col);
            if (idx !== -1) alive.splice(idx,1);
            shotBoard[row][col] = 'X';
            enemyBoard[row][col] = 'X';
            return 'X';
        } else {
            shotBoard[row][col] = 'O';
            enemyBoard[row][col] = 'O';
            return 'O';
        }
    }

    printDual() {
        console.log('\nВаше поле:');
        this.printBoard(this.board1, true);
        console.log('\nПоле противника:');
        this.printBoard(this.shot2, false);
        console.log(`Выстрелов: ${this.shots}, Попаданий: ${this.hits}`);
    }

    playerTurn() {
        return new Promise((resolve) => {
            const ask = () => {
                this.printDual();
                rl.question('Ваш ход (A1): ', (coord) => {
                    if (coord === 'q') process.exit();
                    const pos = this.parseCoord(coord);
                    if (!pos) {
                        console.log('Неверный формат.');
                        ask();
                        return;
                    }
                    const [row, col] = pos;
                    if (this.shot2[row][col] !== '~') {
                        console.log('Сюда уже стреляли.');
                        ask();
                        return;
                    }
                    const res = this.makeShot(row, col, this.alive2, this.shot2, this.board2);
                    this.shots++;
                    if (res === 'X') {
                        this.hits++;
                        console.log('Попадание!');
                        if (this.alive2.length === 0) {
                            console.log('Все корабли противника потоплены! Вы победили!');
                            resolve(1);
                            return;
                        }
                        ask(); // continue
                    } else {
                        console.log('Промах.');
                        resolve(0);
                    }
                });
            };
            ask();
        });
    }

    aiTurn() {
        return new Promise((resolve) => {
            let row, col;
            if (this.difficulty === 'easy') {
                row = Math.floor(Math.random()*SIZE);
                col = Math.floor(Math.random()*SIZE);
            } else {
                let candidates = [];
                for (let r=0; r<SIZE; r++)
                    for (let c=0; c<SIZE; c++)
                        if (this.shot1[r][c] === 'X') {
                            for (const [dr,dc] of [[1,0],[-1,0],[0,1],[0,-1]]) {
                                let nr=r+dr, nc=c+dc;
                                if (nr>=0 && nr<SIZE && nc>=0 && nc<SIZE && this.shot1[nr][nc] === '~')
                                    candidates.push([nr,nc]);
                            }
                        }
                if (candidates.length) {
                    const idx = Math.floor(Math.random()*candidates.length);
                    [row,col] = candidates[idx];
                } else {
                    row = Math.floor(Math.random()*SIZE);
                    col = Math.floor(Math.random()*SIZE);
                }
            }
            if (this.shot1[row][col] === '~') {
                const res = this.makeShot(row, col, this.alive1, this.shot1, this.board1);
                this.shots++;
                if (res === 'X') {
                    this.hits++;
                    console.log(`Компьютер попал! (${this.coordStr(row,col)})`);
                    if (this.alive1.length === 0) {
                        console.log('Все ваши корабли потоплены! Вы проиграли.');
                        resolve(-1);
                        return;
                    }
                    // continue AI turn
                    setTimeout(() => this.aiTurn().then(resolve), 100);
                } else {
                    console.log(`Компьютер промахнулся. (${this.coordStr(row,col)})`);
                    resolve(0);
                }
            } else {
                this.aiTurn().then(resolve);
            }
        });
    }

    async setup() {
        if (this.mode === 'single') {
            if (this.manual) {
                await this.manualPlace(this.board1, this.ships1);
            } else {
                this.placeRandom(this.board1, this.ships1);
            }
            this.placeRandom(this.board2, this.ships2);
            this.alive1 = this.ships1.flat();
            this.alive2 = this.ships2.flat();
        } else {
            console.log('Режим двух игроков пока не реализован.');
        }
    }

    async play() {
        await this.setup();
        if (this.mode === 'single') {
            while (true) {
                if (this.turn === 0) {
                    const res = await this.playerTurn();
                    if (res === 1) break;
                    this.turn = 1;
                } else {
                    const res = await this.aiTurn();
                    if (res === -1) break;
                    this.turn = 0;
                }
            }
        }
    }
}

async function main() {
    let mode = 'single', difficulty = 'easy', manual = false;
    const args = process.argv.slice(2);
    if (args.length && args[0] === 'two') mode = 'two';
    for (let i=1; i<args.length; i++) {
        if (args[i] === '-d' || args[i] === '--difficulty') {
            if (i+1 < args.length) difficulty = args[++i];
        } else if (args[i] === '-p' || args[i] === '--place') {
            manual = true;
        }
    }
    const game = new SeaBattle(mode, difficulty, manual);
    await game.play();
    rl.close();
}

main();

File: seabattle.go
// seabattle.go
package main

import (
	"bufio"
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"time"
)

const SIZE = 10

var colors = map[string]string{
	"reset": "\033[0m",
	"blue": "\033[94m",
	"green": "\033[92m",
	"red": "\033[91m",
	"yellow": "\033[93m",
	"cyan": "\033[96m",
	"gray": "\033[90m",
	"bold": "\033[1m",
}

func colorize(text, color string) string {
	return colors[color] + text + colors["reset"]
}

var SHIP_LENGTHS = []int{4,3,3,2,2,2,1,1,1,1}

type SeaBattle struct {
	board1, board2 [][]byte
	shot1, shot2   [][]byte
	ships1, ships2 [][][2]int
	alive1, alive2 [][2]int
	shots, hits    int
	turn           int
	mode, difficulty string
	manual         bool
	reader         *bufio.Reader
}

func NewSeaBattle(mode, difficulty string, manual bool) *SeaBattle {
	b := &SeaBattle{
		mode: mode,
		difficulty: difficulty,
		manual: manual,
		reader: bufio.NewReader(os.Stdin),
	}
	b.board1 = make([][]byte, SIZE)
	b.board2 = make([][]byte, SIZE)
	b.shot1 = make([][]byte, SIZE)
	b.shot2 = make([][]byte, SIZE)
	for i := 0; i < SIZE; i++ {
		b.board1[i] = make([]byte, SIZE)
		b.board2[i] = make([]byte, SIZE)
		b.shot1[i] = make([]byte, SIZE)
		b.shot2[i] = make([]byte, SIZE)
		for j := 0; j < SIZE; j++ {
			b.board1[i][j] = '~'
			b.board2[i][j] = '~'
			b.shot1[i][j] = '~'
			b.shot2[i][j] = '~'
		}
	}
	return b
}

func (b *SeaBattle) parseCoord(s string) (int, int) {
	if len(s) < 2 { return -1, -1 }
	colChar := s[0]
	if colChar >= 'a' && colChar <= 'z' { colChar -= 32 }
	if colChar < 'A' || colChar > 'J' { return -1, -1 }
	col := int(colChar - 'A')
	row, err := strconv.Atoi(s[1:])
	if err != nil || row < 1 || row > SIZE { return -1, -1 }
	return row-1, col
}

func (b *SeaBattle) coordStr(r, c int) string {
	return string(rune('A'+c)) + strconv.Itoa(r+1)
}

func (b *SeaBattle) canPlace(board [][]byte, row, col, length int, dir byte) bool {
	for i := 0; i < length; i++ {
		var r, c int
		if dir == 'v' {
			r, c = row+i, col
		} else {
			r, c = row, col+i
		}
		if r >= SIZE || c >= SIZE || board[r][c] != '~' { return false }
		for dr := -1; dr <= 1; dr++ {
			for dc := -1; dc <= 1; dc++ {
				nr, nc := r+dr, c+dc
				if nr >= 0 && nr < SIZE && nc >= 0 && nc < SIZE && board[nr][nc] != '~' && board[nr][nc] != '#' {
					if board[nr][nc] == 'S' || board[nr][nc] == 'X' { return false }
				}
			}
		}
	}
	return true
}

func (b *SeaBattle) placeShip(board [][]byte, row, col, length int, dir byte) {
	for i := 0; i < length; i++ {
		var r, c int
		if dir == 'v' {
			r, c = row+i, col
		} else {
			r, c = row, col+i
		}
		board[r][c] = 'S'
	}
}

func (b *SeaBattle) placeRandom(board [][]byte, ships *[][][2]int) {
	rand.Seed(time.Now().UnixNano())
	for _, length := range SHIP_LENGTHS {
		placed := false
		attempts := 0
		for !placed && attempts < 1000 {
			attempts++
			row, col := rand.Intn(SIZE), rand.Intn(SIZE)
			dir := 'h'
			if rand.Intn(2) == 0 { dir = 'v' }
			if b.canPlace(board, row, col, length, byte(dir)) {
				b.placeShip(board, row, col, length, byte(dir))
				var coords [][2]int
				for i := 0; i < length; i++ {
					var r, c int
					if dir == 'v' {
						r, c = row+i, col
					} else {
						r, c = row, col+i
					}
					coords = append(coords, [2]int{r, c})
				}
				*ships = append(*ships, coords)
				placed = true
			}
		}
		if !placed {
			// reset and retry
			for i := range board {
				for j := range board[i] {
					board[i][j] = '~'
				}
			}
			*ships = [][][2]int{}
			b.placeRandom(board, ships)
			return
		}
	}
}

func (b *SeaBattle) manualPlace(board [][]byte, ships *[][][2]int) {
	fmt.Println("Ручная расстановка кораблей.")
	for _, length := range SHIP_LENGTHS {
		placed := false
		for !placed {
			fmt.Printf("Разместите корабль длины %d\n", length)
			b.printBoard(board, true)
			fmt.Print("Введите координату (A1): ")
			coord, _ := b.reader.ReadString('\n')
			coord = strings.TrimSpace(coord)
			if coord == "q" { os.Exit(0) }
			fmt.Print("Введите направление (h/v): ")
			dir, _ := b.reader.ReadString('\n')
			dir = strings.TrimSpace(dir)
			if len(dir) != 1 || (dir != "h" && dir != "v") {
				fmt.Println("Неверный ввод.")
				continue
			}
			row, col := b.parseCoord(coord)
			if row == -1 {
				fmt.Println("Неверные координаты.")
				continue
			}
			if b.canPlace(board, row, col, length, dir[0]) {
				b.placeShip(board, row, col, length, dir[0])
				var coords [][2]int
				for i := 0; i < length; i++ {
					var r, c int
					if dir[0] == 'v' {
						r, c = row+i, col
					} else {
						r, c = row, col+i
					}
					coords = append(coords, [2]int{r, c})
				}
				*ships = append(*ships, coords)
				placed = true
			} else {
				fmt.Println("Нельзя разместить здесь.")
			}
		}
	}
}

func (b *SeaBattle) printBoard(board [][]byte, showShips bool) {
	fmt.Print("   ")
	for i := 0; i < SIZE; i++ {
		fmt.Printf("%c ", 'A'+i)
	}
	fmt.Println()
	for r := 0; r < SIZE; r++ {
		fmt.Printf("%2d ", r+1)
		for c := 0; c < SIZE; c++ {
			cell := board[r][c]
			if cell == '~' {
				fmt.Print(colorize("· ", "gray"))
			} else if cell == 'S' && showShips {
				fmt.Print(colorize("S ", "green"))
			} else if cell == 'X' {
				fmt.Print(colorize("X ", "red"))
			} else if cell == 'O' {
				fmt.Print(colorize("O ", "yellow"))
			} else {
				fmt.Printf("%c ", cell)
			}
		}
		fmt.Println()
	}
}

func (b *SeaBattle) makeShot(row, col int, alive *[][2]int, shotBoard, enemyBoard [][]byte) byte {
	if enemyBoard[row][col] == 'S' {
		// удалить из alive
		newAlive := [][2]int{}
		for _, p := range *alive {
			if !(p[0] == row && p[1] == col) {
				newAlive = append(newAlive, p)
			}
		}
		*alive = newAlive
		shotBoard[row][col] = 'X'
		enemyBoard[row][col] = 'X'
		return 'X'
	} else {
		shotBoard[row][col] = 'O'
		enemyBoard[row][col] = 'O'
		return 'O'
	}
}

func (b *SeaBattle) printDual() {
	fmt.Println("\nВаше поле:")
	b.printBoard(b.board1, true)
	fmt.Println("\nПоле противника:")
	b.printBoard(b.shot2, false)
	fmt.Printf("Выстрелов: %d, Попаданий: %d\n", b.shots, b.hits)
}

func (b *SeaBattle) playerTurn() int {
	for {
		b.printDual()
		fmt.Print("Ваш ход (A1): ")
		coord, _ := b.reader.ReadString('\n')
		coord = strings.TrimSpace(coord)
		if coord == "q" { os.Exit(0) }
		row, col := b.parseCoord(coord)
		if row == -1 {
			fmt.Println("Неверный формат.")
			continue
		}
		if b.shot2[row][col] != '~' {
			fmt.Println("Сюда уже стреляли.")
			continue
		}
		res := b.makeShot(row, col, &b.alive2, b.shot2, b.board2)
		b.shots++
		if res == 'X' {
			b.hits++
			fmt.Println("Попадание!")
			if len(b.alive2) == 0 {
				fmt.Println("Все корабли противника потоплены! Вы победили!")
				return 1
			}
			continue
		} else {
			fmt.Println("Промах.")
			break
		}
	}
	return 0
}

func (b *SeaBattle) aiTurn() int {
	rand.Seed(time.Now().UnixNano())
	for {
		var row, col int
		if b.difficulty == "easy" {
			row, col = rand.Intn(SIZE), rand.Intn(SIZE)
		} else {
			var candidates [][2]int
			for r := 0; r < SIZE; r++ {
				for c := 0; c < SIZE; c++ {
					if b.shot1[r][c] == 'X' {
						for _, d := range [][2]int{{1,0},{-1,0},{0,1},{0,-1}} {
							nr, nc := r+d[0], c+d[1]
							if nr >= 0 && nr < SIZE && nc >= 0 && nc < SIZE && b.shot1[nr][nc] == '~' {
								candidates = append(candidates, [2]int{nr, nc})
							}
						}
					}
				}
			}
			if len(candidates) > 0 {
				idx := rand.Intn(len(candidates))
				row, col = candidates[idx][0], candidates[idx][1]
			} else {
				row, col = rand.Intn(SIZE), rand.Intn(SIZE)
			}
		}
		if b.shot1[row][col] == '~' {
			res := b.makeShot(row, col, &b.alive1, b.shot1, b.board1)
			b.shots++
			if res == 'X' {
				b.hits++
				fmt.Printf("Компьютер попал! (%s)\n", b.coordStr(row, col))
				if len(b.alive1) == 0 {
					fmt.Println("Все ваши корабли потоплены! Вы проиграли.")
					return -1
				}
				continue
			} else {
				fmt.Printf("Компьютер промахнулся. (%s)\n", b.coordStr(row, col))
				break
			}
		}
	}
	return 0
}

func (b *SeaBattle) setup() {
	if b.mode == "single" {
		if b.manual {
			b.manualPlace(b.board1, &b.ships1)
		} else {
			b.placeRandom(b.board1, &b.ships1)
		}
		b.placeRandom(b.board2, &b.ships2)
		for _, ship := range b.ships1 {
			for _, p := range ship {
				b.alive1 = append(b.alive1, p)
			}
		}
		for _, ship := range b.ships2 {
			for _, p := range ship {
				b.alive2 = append(b.alive2, p)
			}
		}
	} else {
		fmt.Println("Режим двух игроков пока не реализован.")
	}
}

func (b *SeaBattle) play() {
	b.setup()
	if b.mode == "single" {
		for {
			if b.turn == 0 {
				res := b.playerTurn()
				if res == 1 { break }
				b.turn = 1
			} else {
				res := b.aiTurn()
				if res == -1 { break }
				b.turn = 0
			}
		}
	}
}

func main() {
	mode := "single"
	difficulty := "easy"
	manual := false
	if len(os.Args) > 1 {
		if os.Args[1] == "two" { mode = "two" }
	}
	for i := 2; i < len(os.Args); i++ {
		if os.Args[i] == "-d" || os.Args[i] == "--difficulty" {
			if i+1 < len(os.Args) {
				difficulty = os.Args[i+1]
				i++
			}
		} else if os.Args[i] == "-p" || os.Args[i] == "--place" {
			manual = true
		}
	}
	game := NewSeaBattle(mode, difficulty, manual)
	game.play()
}
